src/arrow1/reader_async.rs

```diff
@@ -1,15 +1,670 @@
-use parquet::file::metadata::FileMetaData;
+use parquet::file::footer::{decode_footer, decode_metadata};
+use parquet::file::metadata::{FileMetaData, ParquetMetaData};
 use range_reader::RangedAsyncReader;
+use reqwest::Client;
+
+use async_trait::async_trait;
+
+#[async_trait]
+trait SharedIO<T: AsyncFileReader + Unpin + Sync + Clone + 'static> {
+    fn generate_builder(
+        reader: &T,
+        meta: &ArrowReaderMetadata,
+        batch_size: &usize,
+        projection_mask: &Option<ProjectionMask>,
+    ) -> ParquetRecordBatchStreamBuilder<T> {
+        let builder =
+            ParquetRecordBatchStreamBuilder::new_with_metadata(reader.clone(), meta.clone())
+                .with_batch_size(*batch_size)
+                .with_projection(
+                    projection_mask
+                        .as_ref()
+                        .unwrap_or(&ProjectionMask::all())
+                        .clone(),
+                );
+        builder
+    }
+
+    async fn inner_read_row_group(
+        &self,
+        reader: &T,
+        meta: &ArrowReaderMetadata,
+        batch_size: &usize,
+        projection_mask: &Option<ProjectionMask>,
+        i: usize,
+    ) -> Result<Table> {
+        let builder = Self::generate_builder(reader, meta, batch_size, projection_mask);
+        let stream = builder.with_row_groups(vec![i]).build()?;
+        let results = stream.try_collect::<Vec<_>>().await.unwrap();
+
+        // NOTE: This is not only one batch by default due to arrow-rs's default rechunking.
+        // assert_eq!(results.len(), 1, "Expected one record batch");
+        // Ok(RecordBatch::new(results.pop().unwrap()))
+        Ok(Table::new(results))
+    }
+
+    async fn inner_stream(
+        &self,
+        concurrency: Option<usize>,
+        meta: &ArrowReaderMetadata,
+        reader: &T,
+        batch_size: &usize,
+        projection_mask: &Option<ProjectionMask>,
+    ) -> WasmResult<wasm_streams::readable::sys::ReadableStream> {
+        use futures::StreamExt;
+        let concurrency = concurrency.unwrap_or(1);
+        let meta = meta.clone();
+        let reader = reader.clone();
+        let batch_size = *batch_size;
+        let num_row_groups = meta.metadata().num_row_groups();
+        let projection_mask = projection_mask.clone();
+        let buffered_stream = stream::iter((0..num_row_groups).map(move |i| {
+            let builder = Self::generate_builder(&reader, &meta, &batch_size, &projection_mask)
+                .with_row_groups(vec![i]);
+            builder.build().unwrap().try_collect::<Vec<_>>()
+        }))
+        .buffered(concurrency);
+        let out_stream = buffered_stream.flat_map(|maybe_record_batches| {
+            stream::iter(maybe_record_batches.unwrap())
+                .map(|record_batch| Ok(RecordBatch::new(record_batch).into()))
+        });
+        Ok(wasm_streams::ReadableStream::from_stream(out_stream).into_raw())
+    }
+}
+
+#[wasm_bindgen]
+pub struct AsyncParquetFile {
```

Comment:
fwiw in a future PR I think in the future I'd like to have `AsyncParquetDataset` as well, which could just be a `HashMap<String, AsyncParquetFile>`, and would allow managing multi-file remote Parquet stores